src/cookbooks/client_guidelines.md

```diff
@@ -0,0 +1,97 @@
+# Client Implementation Guidelines
+## Feature Gates
+
+The preferred method for testing experimental features is to __land in-tree__ and ship behind a __feature gate__. New features go through the standard Firefox review, testing, and deployment processes, and are then enabled experimentally in the field using these gates.
+A new Feature Gate library for Firefox Desktop is now available. Read [the documentation][feature-gate-docs] for it to get started.
+
+New features should use this library or an equivalent feature-flagging system as appropriate for their platform.
+
+Historically, Firefox Desktop has used the Preferences system to optionally activate features.
+
+__In order for features to be activated experimentally, these rules must be observed__:
+
+- It must be possible to enable/disable the feature at runtime, via a pref change.
+- Use `default branch`, not `user branch` prefs.
+  - For instance, never use `Services.prefs.prefHasUserValue()`.
+- Prefs should be set by default in `firefox.js`
+
+If your feature cannot abide by one or more of these rules (for instance, it needs to run at startup and/or cannot be toggled at runtime) then experimental preferences can be set on the user branch. This isn't preferred, but will work with features that require a restart. This kind of experiment will require additional engineering support.
```

Comment:
This section (from "In order for features" to "additional engineering support") talk about preferences, but the surrounding text talks about feature gates. I think these rules only apply to pref based features, not feature gate based features.

Reply:
Good point, thanks... I'm struggling a bit to separate these, and make it clear that gates are preferred but prefs are going to be required for a while still (at least until we have sync and async APIs and support for native code)